### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by egbertbouman**

### CODE REVIEW

**Comment by qstokkink** [comment]
We should let the Python 3 tests on Linux run all the tests now I guess.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by devos50** [comment]
I will start to review this PR 👍

**Comment by devos50** [comment]
I made a job that runs all tests with Python 3 in parallel, see [here](https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_linux_py3_all/). It seems that all tests are passing, see this [build](https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_linux_py3_all/8/).

**Comment by qstokkink** [comment]
False positive on the `External import checker`, `tkinter` is no longer shipped in Python by default (we should maybe look into this later).

File: Tribler/Test/Core/Modules/test_process_checker.py [hunk 1]
```diff
@@ -28,7 +32,7 @@ def setUp(self):
 
     def create_lock_file_with_pid(self, pid):
         with open(os.path.join(self.state_dir, LOCK_FILE_NAME), 'wb') as lock_file:
-            lock_file.write(str(pid))
+            lock_file.write(text_type(pid).encode())
```
Comment (devos50) [inline Tribler/Test/Core/Modules/test_process_checker.py]: Why not `ensure_binary` (or just open the file in `w` mode)?

File: Tribler/Test/Core/Video/test_vod.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def state_callback(self, ds):
             stream = VODFile(open(download.get_content_dest(), 'rb'), download)
 
             # Read last piece
-            lastpieceoff = ((self.contentlen - 1) / self.piecelen) * self.piecelen
+            lastpieceoff = ((self.contentlen - 1) // self.piecelen) * self.piecelen
```
Comment (devos50) [inline Tribler/Test/Core/Video/test_vod.py]: Did you import `division` ?

File: Tribler/Test/util/Tracker/UDPTracker.py [hunk 3]
```diff
@@ -42,11 +45,11 @@ def datagramReceived(self, response, host_and_port):
                 self.send_error(host, port, "no infohash")
                 return
 
-            num_infohashes = (len(response) - 16) / LENGTH_INFOHASH
+            num_infohashes = (len(response) - 16) // LENGTH_INFOHASH
```
Comment (devos50) [inline Tribler/Test/util/Tracker/UDPTracker.py]: Import `division`?

File: Tribler/community/triblertunnel/community.py [hunk 4]
```diff
@@ -301,7 +301,7 @@ def on_relay_balance_response_cell(self, source_address, data, _):
                                BalanceResponsePayload.from_half_block(block, cache.to_circuit_id))
 
     def readd_bittorrent_peers(self):
-        for torrent, peers in self.bittorrent_peers.items():
+        for torrent, peers in list(self.bittorrent_peers.items()):
```
Comment (devos50) [inline Tribler/community/triblertunnel/community.py]: See my prior comments about iteration.

File: TriblerGUI/dialogs/startdownloaddialog.py [hunk 5]
```diff
@@ -183,7 +183,7 @@ def on_received_metainfo(self, metainfo):
         self.dialog_widget.files_list_view.clear()
         for filename in files:
             item = DownloadFileTreeWidgetItem(self.dialog_widget.files_list_view)
-            item.setText(0, '/'.join(filename['path']).encode('raw_unicode_escape'))
+            item.setText(0, '/'.join(filename['path']).encode('raw_unicode_escape').decode('utf-8'))
```
Comment (devos50) [inline TriblerGUI/dialogs/startdownloaddialog.py]: Can this be simplified?

File: TriblerGUI/widgets/downloadspage.py [hunk 6]
```diff
@@ -169,7 +169,7 @@ def on_received_downloads(self, downloads):
             self.window().downloads_list.setItemWidget(item, 2, item.bar_container)
 
         # Check whether there are download that should be removed
-        for infohash, item in self.download_widgets.items():
+        for infohash, item in list(self.download_widgets.items()):
```
Comment (devos50) [inline TriblerGUI/widgets/downloadspage.py]: See my prior comments on using `lists`.

**Comment by devos50** [comment]
@egbertbouman  there are some fixable pylint errors still.

File: Tribler/Test/util/Tracker/UDPTracker.py (hunk 3 above)
Comment (qstokkink) [reply Tribler/Test/util/Tracker/UDPTracker.py]: ⏫ This is required.

File: Tribler/Core/CreditMining/CreditMiningManager.py [hunk 7]
```diff
@@ -189,7 +189,7 @@ def remove_source(self, source_str):
             self._logger.info('Removed source %s', source_str)
 
             deferreds = []
-            for infohash, torrent in self.torrents.items():
+            for infohash, torrent in list(self.torrents.items()):
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/CreditMining/CreditMiningManager.py]: Dict changes while iteration, so I'm iterating over a copy.

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 8]
```diff
@@ -54,7 +50,7 @@ def __init__(self, f, d):
         self._download = d
 
         pieces = self._download.tdef.get_pieces()
-        self.pieces = [pieces[x:x + 20] for x in xrange(0, len(pieces), 20)]
+        self.pieces = [pieces[x:x + 20] for x in list(range(0, len(pieces), 20))]
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: Probably not, I'll remove it.

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 9]
```diff
@@ -298,12 +281,13 @@ def network_create_engine_wrapper(self, pstate, checkpoint_disabled=False, share
...
-                        resume_data["save_path"] = six.text_type(
-                            os.path.join(self.state_dir, ensure_unicode(resume_data["save_path"], 'utf-8')))
+                    if b"save_path" in resume_data and not os.path.isabs(resume_data[b"save_path"]):
+                        resume_data[b"save_path"] = text_type(
+                            os.path.join(self.state_dir, ensure_unicode(resume_data[b"save_path"], 'utf-8')))
                     atp["resume_data"] = lt.bencode(resume_data)
             else:
-                atp["url"] = self.tdef.get_url() or "magnet:?xt=urn:btih:" + hexlify(self.tdef.get_infohash())
+                atp["url"] = self.tdef.get_url() or \
+                             "magnet:?xt=urn:btih:" + hexlify(self.tdef.get_infohash())
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: This is our own `hexlify`, which return unicode

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 10]
```diff
@@ -644,15 +630,14 @@ def on_torrent_checked_alert(self, alert):
     @checkHandleAndSynchronize()
     def on_torrent_finished_alert(self, alert):
         self.update_lt_status(self.handle.status())
-        if self.finished_deferred_already_called:
+        if self.finished_deferred.called:
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: It works on Twisted 16.6.0, which is from 2016. It's still in Twisted today, so I figured it's safe to use.

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 11]
```diff
@@ -829,17 +814,15 @@ def set_moreinfo_stats(self, enable):
...
                 return self.get_byte_progress(
                     [(self.get_vod_fileindex(), self.vod_seekpos, self.vod_seekpos + self.prebuffsize),
                      (self.get_vod_fileindex(), -self.endbuffsize - 1, -1)], consecutive=consecutive)
-            else:
-                return self.get_byte_progress(
-                    [(self.get_vod_fileindex(), self.vod_seekpos, self.vod_seekpos + self.prebuffsize)],
-                    consecutive=consecutive)
-        else:
-            return 0.0
+            return self.get_byte_progress(
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: It's a Pylint fix.

File: Tribler/Core/Modules/MetadataStore/OrmBindings/channel_metadata.py [hunk 12]
```diff
@@ -68,8 +66,7 @@ def entries_to_chunk(metadata_list, chunk_size, start_index=0):
         offset = len(header)
         out_list = [header]  # LZ4 header
         for index, metadata in enumerate(metadata_list[start_index:], start_index):
-            blob = c.compress(
-                ''.join(metadata.serialized_delete() if metadata.status == TODELETE else metadata.serialized()))
+            blob = c.compress((metadata.serialized_delete() if metadata.status == TODELETE else metadata.serialized()))
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Modules/MetadataStore/OrmBindings/channel_metadata.py]: It's a join with only 1 argument, so it didn't really do anything.

File: Tribler/Core/TorrentDef.py [hunk 13]
```diff
@@ -494,7 +481,7 @@ class TorrentDefNoMetainfo(object):
     """
 
     def __init__(self, infohash, name, url=None):
-        assert isinstance(infohash, six.binary_type), "INFOHASH has invalid type: %s" % type(infohash)
+        #assert isinstance(infohash, six.binary_type), "INFOHASH has invalid type: %s" % type(infohash)
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/TorrentDef.py]: Oops, I missed that one.

File: Tribler/Core/TorrentDef.py [hunk 14]
```diff
@@ -522,10 +509,10 @@ def get_name_utf8(self):
         """
         Not all names are utf-8, attempt to construct it as utf-8 anyway.
         """
-        return escape_as_utf8(self.get_name())
+        return escape_as_utf8(six.ensure_binary(self.get_name()))
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/TorrentDef.py]: Will do.

File: Tribler/Core/Utilities/maketorrent.py [hunk 15]
```diff
@@ -23,47 +23,45 @@ def pathlist2filename(pathlist):
...
-
-            if length > 0 and (not selectedfiles or (selectedfiles and filename in selectedfiles)):
-                pieces_range = (offset_to_piece(offset, piecesize, False), offset_to_piece(offset + length, piecesize),
-                                (offset - offset_to_piece(offset, piecesize, False) * piecesize), filename)
-                filepieceranges.append(pieces_range)
-                total += length
-            offset += length
-        return total, filepieceranges
+        return metainfo[b'info'][b'length'], None
+    # multi-file torrent
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Utilities/maketorrent.py]: If we want the keep the else statements, I guess we should change the Pylint config.

File: Tribler/Core/Utilities/random_utils.py [hunk 16]
```diff
@@ -9,7 +13,7 @@ def random_string(size=6, chars=string.ascii_uppercase + string.digits):
 
 def random_infohash():
     """ Generates a random torrent infohash string """
-    return ''.join(random.choice('0123456789abcdef') for _ in range(20))
+    return hexlify(os.urandom(20))
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Core/Utilities/random_utils.py]: Note that the logic does not change here. The stuff that uses it requires a string

File: Tribler/Test/Community/gigachannel/test_community.py [hunk 17]
```diff
@@ -195,12 +195,13 @@ def mock_notify(overlay, args):
             self.assertEqual(len(torrents), 5)
 
             # Only non-legacy FFA torrents should be sent on search
+            empty_pk = database_blob(b"")
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Test/Community/gigachannel/test_community.py]: Sure.

File: Tribler/Test/Core/Modules/MetadataStore/test_channel_metadata.py [hunk 18]
```diff
@@ -226,7 +226,7 @@ def test_restore_torrent_in_channel(self):
         # Check update of torrent properties from a new tdef
         md.status = TODELETE
         new_tracker_address = u'http://tribler.org/announce'
-        tdef.torrent_parameters['announce'] = new_tracker_address
+        tdef.torrent_parameters[b'announce'] = new_tracker_address.encode('utf-8')
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Test/Core/Modules/MetadataStore/test_channel_metadata.py]: This variable is also used elsewhere, and there it needs to be unicode.

File: Tribler/Test/Core/Modules/RestApi/test_downloads_endpoint.py [hunk 19]
```diff
@@ -310,13 +311,13 @@ def verify_removed(_):
...
         video_tdef, _ = self.create_local_torrent(os.path.join(TESTS_DATA_DIR, 'video.avi'))
-        self.session.start_download_from_tdef(video_tdef, DownloadStartupConfig())
+        self.session.start_download_from_tdef(video_tdef, DownloadConfig())
         infohash = get_hex_infohash(video_tdef)
 
         request_deferred = self.do_request('downloads/%s' % infohash, post_data={"remove_data": True},
                                            expected_code=200, request_type='DELETE',
-                                           expected_json={"removed": True,
-                                                          "infohash": "c9a19e7fe5d9a6c106d6ea3c01746ac88ca3c7a5"})
+                                           expected_json={u"removed": True,
```
[the remark this answers is not in the document]
Comment (egbertbouman) [reply Tribler/Test/Core/Modules/RestApi/test_downloads_endpoint.py]: We don't. I'll remove it.

File: Tribler/Test/util/Tracker/UDPTracker.py (hunk 3 above)
Comment (egbertbouman) [reply Tribler/Test/util/Tracker/UDPTracker.py]: Why? The `//` operator refers to integer division in both Python 2 and 3.

**Comment by devos50** [comment]
Reminder to myself: as soon as we release 7.3.0, I suggest we start setting up Python 3 environments on our testing machines and builders. This includes a py3-compatible libtorrent.

**Pull request merged**